### PR TITLE
Extension Trait for Tasks and Events

### PR DESCRIPTION
**Pull request by jamesmunns**
~~Note: This is currently only for the nrf52832, but I wanted to get your feedback before updating for all targets.~~

Implemented for all HAL targets

### CODE REVIEW

File: nrf-hal-common/src/ppi/event_nrf9160.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+use crate::ppi::Event;
+
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/event_nrf9160.rs]: Hmm, the `PPI` interface on the nrf9160 is quite different and not supported by this implementation, so this file won't be used, maybe we should remove it then ?

File: nrf-hal-common/src/ppi/mod.rs [hunk 2]
```diff
@@ -10,15 +10,54 @@
...
+    } else if #[cfg(feature = "52832")] {
+        mod event_nrf52832;
+        mod task_nrf52832;
+    } else if #[cfg(feature = "52833")] {
+        mod event_nrf52833;
+        mod task_nrf52833;
+    } else if #[cfg(feature = "52840")] {
+        mod event_nrf52840;
+        mod task_nrf52840;
+    } else if #[cfg(feature = "9160")] {
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs]: Not really supported.

File: nrf-hal-common/src/ppi/mod.rs [hunk 3]
```diff
@@ -31,18 +70,18 @@ pub trait Ppi {
     #[cfg(not(feature = "51"))]
     /// Sets the fork task that must be triggered when the configured event occurs. The user must
     /// provide the address of the task.
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs]:
```suggestion
    /// provide a reference to the task.
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 4]
```diff
@@ -31,18 +70,18 @@ pub trait Ppi {
...
-    fn set_fork_task_endpoint(&mut self, addr: u32);
+    fn set_fork_task_endpoint<T: Task>(&mut self, task: &T);
 }
 
 /// Traits that extends the [Ppi](trait.Ppi.html) trait, marking a channel as fully configurable.
 pub trait ConfigurablePpi {
     /// Sets the task that must be triggered when the configured event occurs. The user must provide
     /// the address of the task.
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs]:
```suggestion
    /// a reference to the task.
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 5]
```diff
@@ -31,18 +70,18 @@ pub trait Ppi {
...
-    fn set_task_endpoint(&mut self, addr: u32);
+    fn set_task_endpoint<T: Task>(&mut self, task: &T);
 
     /// Sets the event that will trigger the chosen task(s). The user must provide the address of
     /// the event.
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs]:
```suggestion
    /// Sets the event that will trigger the chosen task(s). The user must provide a reference to
    /// the event.
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 6]
```diff
@@ -59,23 +98,29 @@ impl<P: Channel> Ppi for P {
     }
 
     #[cfg(not(feature = "51"))]
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs:101]:
```suggestion
    #[cfg(not(feature = "51"))]
    #[inline(always)]
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 7]
```diff
@@ -59,23 +98,29 @@ impl<P: Channel> Ppi for P {
...
+            .tep
+            .write(|w| unsafe { w.bits(task.task_addr().0) });
     }
 }
 
 // All unsafe `ptr` calls only uses registers atomically, and only changes the resources owned by
 // the type (guaranteed by the abstraction)
 impl<P: Channel + NotFixed> ConfigurablePpi for P {
-    fn set_task_endpoint(&mut self, addr: u32) {
+    fn set_task_endpoint<T: Task>(&mut self, task: &T) {
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs:115]:
```suggestion
    #[inline(always)]
    fn set_task_endpoint<T: Task>(&mut self, task: &T) {
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 8]
```diff
@@ -59,23 +98,29 @@ impl<P: Channel> Ppi for P {
...
         let regs = unsafe { &*PPI::ptr() };
-        regs.ch[P::CH].tep.write(|w| unsafe { w.bits(addr) });
+        regs.ch[P::CH]
+            .tep
+            .write(|w| unsafe { w.bits(task.task_addr().0) });
     }
 
-    fn set_event_endpoint(&mut self, addr: u32) {
+    fn set_event_endpoint<E: Event>(&mut self, event: &E) {
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs:123]:
```suggestion
    #[inline(always)]
    fn set_event_endpoint<E: Event>(&mut self, event: &E) {
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 9]
```diff
@@ -10,15 +10,54 @@
...
 
 mod sealed {
+    use super::{EventAddr, TaskAddr};
+
     pub trait Channel {
         const CH: usize;
     }
 
+    pub trait Task {
+        fn task_addr(&self) -> TaskAddr {
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs:43]:
```suggestion
        #[inline(always)]
        fn task_addr(&self) -> TaskAddr {
```

File: nrf-hal-common/src/ppi/mod.rs [hunk 10]
```diff
@@ -10,15 +10,54 @@
...
+            TaskAddr(self as *const _ as *const u32 as u32)
+        }
+    }
+    pub trait Event {
+        fn event_addr(&self) -> EventAddr {
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/mod.rs:49]:
```suggestion
        #[inline(always)]
        fn event_addr(&self) -> EventAddr {
```

File: nrf-hal-common/src/ppi/task_nrf9160.rs [hunk 11]
```diff
@@ -0,0 +1,65 @@
+use crate::ppi::Task;
+
```
Comment (thalesfragoso) [inline nrf-hal-common/src/ppi/task_nrf9160.rs]: The nRF9160 is not supported by this implementation.

**Review by thalesfragoso — CHANGES_REQUESTED** [review]
Thanks for the nice work, I added some suggestions, could you also put the `#[inline(always)]` in the places I forgot ? Thanks.

**Comment by jamesmunns** [comment]
@thalesfragoso addressed your changes!

**Review by thalesfragoso — APPROVED** [review]
Thank you!

**Pull request merged**